File: Queue/Circular_Queue.cpp
#include<iostream>
using namespace std;
struct Circular_Queue
{
    int queue[5];
    int front = -1;
    int rear = -1;
};
Circular_Queue q;
void insertion(int value,int size)
{
    if((q.front == 0  && q.rear == size-1) || (q.front == q.rear+1))
    {
        cout<<"Overflow";
        return;
    }
    else if (q.front == -1 && q.rear == -1)

    {
        q.front++;
        q.rear++;
        q.queue[q.rear] = value;
    }
    else if (q.front!=0 && q.rear==size-1)
    {
        q.rear=0;
        q.queue[q.rear] = value;
    }
    else{
        q.rear++;
        q.queue[q.rear] = value;
    }
    
    
}
void deletion(int size)
{
    if(q.front == -1 && q.rear == -1)
    {
        cout<<"Underflow";
        return;
    }
    else if (q.front == q.rear)
    {
        q.front = -1;
        q.rear = -1;
    }
    else if (q.front == size-1)
    {
        q.front = 0;
    }
    else{
        int tem = q.queue[q.front];
        q.front++;
    }
    
    
}

int main()
{
    insertion(1,5);
    insertion(2,5);
    insertion(3,5);
    insertion(4,5);
    insertion(5,5);
    deletion(5);
    deletion(5);
    deletion(5);
    insertion(6,5);
    deletion(5);
    if (q.front <= q.rear)
    {
        for (int i = q.front; i <= q.rear; i++)
        {
            cout << q.queue[i] << " ";
        }
    }
    else
    {
        for (int i = q.front; i < 5; i++)
        {
            cout << q.queue[i] << " ";
        }
        for (int i = 0; i <= q.rear; i++)
        {
            cout << q.queue[i] << " ";
        }
    }
    cout << endl;

    return 0;
}
    

File: Queue/src/CircularQueue.java
import java.util.Objects;

public class CircularQueue {
    static class Queue
    {
        static int arr[];
        static int size;
        static int rear = -1;
        static int front = -1;
        Queue(int n)
        {
            arr = new int[n];
            this.size = n;
        }
        public static boolean isEmpty()
        {
            return rear == -1 && front == -1;
        }
        public static boolean isFull()
        {
            return (rear+1)%size == front;
        }

        public static void add(int data)
        {
            if(isFull())
            {
                System.out.println("Queue is full");
                return;
            }
            if(front == -1) //1st element add
            {
                front = 0;
            }
            rear = (rear+1)%size;
            arr[rear] = data;

        }

        public static int remove()
        {
            if(isEmpty())
            {
                System.out.println("Queue empty");
                return -1;
            }
            int result = arr[front];
            //single element condition
            if(front == rear)
            {
                front = rear = -1;
            }
            else
            {
            front = (front+1)%size;
            }

            return result;

        }
        public static int peek()
        {
            if(isEmpty())
            {
                System.out.println("Empty queue");
                return -1;
            }
            return arr[front];

        }
    }

    public static void main(String[] args) {
        Queue q = new Queue(5);
        q.add(1);
        q.add(2);
        q.add(3);
        q.add(4);
        q.add(5);
        System.out.println(q.remove());
        q.add(6);
        System.out.println(q.remove());

        while (!q.isEmpty())
        {
            System.out.println(q.peek());
            q.remove();
        }


    }
}
